src/FullTextSearchHelpers.ts

```diff
@@ -0,0 +1,118 @@
+    'ȉ': 'i', 'Ȋ': 'I', 'ȋ': 'i', 'Ȍ': 'O', 'ȍ': 'o', 'Ȏ': 'O', 'ȏ': 'o', 'Ȑ': 'R', 'ȑ': 'r', 'Ȓ': 'R',
+    'ȓ': 'r', 'Ȕ': 'U', 'ȕ': 'u', 'Ȗ': 'U', 'ȗ': 'u', 'Ș': 'S', 'ș': 's', 'Ț': 'T', 'ț': 't', 'Ȟ': 'H',
+    'ȟ': 'h', 'Ȧ': 'A', 'ȧ': 'a', 'Ȩ': 'E', 'ȩ': 'e', 'Ȫ': 'O', 'ȫ': 'o', 'Ȭ': 'O', 'ȭ': 'o', 'Ȯ': 'O',
+    'ȯ': 'o', 'Ȱ': 'O', 'ȱ': 'o', 'Ȳ': 'Y', 'ȳ': 'y', 'ʰ': 'h', 'ʲ': 'j', 'ʳ': 'r', 'ʷ': 'w', 'ʸ': 'y',
+    'ˡ': 'l', 'ˢ': 's', 'ˣ': 'x'
+};
+const _minLead = parseInt('0xD800', 16);
+const _maxLead = parseInt('0xDBFF', 16);
+
+export function normalizeString(input: string): string {
+    // Constants for leading byte-range.
+    let currentChar = '';
+    let output = '';
+
+    if (!input) {
+        return input;
+    }
+
+    for (var i = 0; i < input.length; i++) {
+        currentChar += input.charAt(i);
+        const code = input.charCodeAt(i);
+        if (code < _minLead || code > _maxLead) {
+            output += _charMap[currentChar] ? _charMap[currentChar] : currentChar;
+            currentChar = '';
+        }
+    }
+
+    return output;
+}
+
+export function breakWords(rawString: string): string[] {
+    // Figure out how to do this in a localized fashion
+    return rawString.split(' ');
+}
+
+export function breakAndNormalizeSearchPhrase(phrase: string): string[] {
+    return _.uniq(_.map(breakWords(phrase), word => normalizeSearchTerm(word)));
+}
+
+export function normalizeSearchTerm(term: string): string {
+    return normalizeString(term).toLowerCase();
+}
+
+export function getFullTextIndexWordsForItem(keyPath: string, item: any): string[] {
+    const rawString = NoSqlProviderUtils.getValueForSingleKeypath(item, keyPath);
+
+    return breakAndNormalizeSearchPhrase(rawString);
+}
+
+export abstract class DbIndexFTSFromRangeQueries implements NoSqlProvider.DbIndex {
+    constructor(protected _primaryKeyPath: string | string[]) {
+        // NOP
+    }
+
+    fullTextSearch<T>(searchPhrase: string): SyncTasks.Promise<T[]> {
+        const promises = _.map(breakAndNormalizeSearchPhrase(searchPhrase), term => {
+            const upperEnd = term.substr(0, term.length - 1) + String.fromCharCode(term.charCodeAt(term.length - 1) + 1);
+            return this.getRange(term, upperEnd, false, true);
+        });
+        return SyncTasks.all(promises).then(results => {
+            if (results.length === 1) {
+                return results[0];
+            }
+
+            // Only return terms that show up in all of the results sets.
+            // The @types for _.intersectionBy is wrong and needs fixing, so this will hack around that for now...
+            return (_.intersectionBy as any)(...results, item =>
+                NoSqlProviderUtils.getSerializedKeyForKeypath(item, this._primaryKeyPath));
+        });
+    }
+
+    abstract getAll<T>(reverse?: boolean, limit?: number, offset?: number): SyncTasks.Promise<T[]>;
```

Comment:
TypeScript does not require an abstract class to redefine abstract methods: it knows the implementation will need them. Or are these added?